snap/gadget.go

```diff
@@ -234,6 +256,231 @@ func ReadGadgetInfo(info *Info, classic bool) (*GadgetInfo, error) {
+	}
+	if err := validateRole(vs, vol); err != nil {
+		var what string
+		if vs.Role != "" {
+			what = fmt.Sprintf("role %q", vs.Role)
+		} else {
+			what = fmt.Sprintf("implicit role %q", vs.Type)
+		}
+		return fmt.Errorf("invalid %s: %v", what, err)
+	}
+	if vs.Filesystem != "" && !strutil.ListContains([]string{"ext4", "vfat", "none"}, vs.Filesystem) {
+		return fmt.Errorf("invalid filesystem %q", vs.Filesystem)
+	}
+
+	var contentChecker func(*VolumeContent) error
+
+	if vs.IsBare() {
+		contentChecker = validateBareContent
+	} else {
+		contentChecker = validateFilesystemContent
+	}
+	for i, c := range vs.Content {
+		if err := contentChecker(&c); err != nil {
+			return fmt.Errorf("invalid content #%v: %v", i, err)
+		}
+	}
+
+	if err := validateStructureUpdate(&vs.Update, vs); err != nil {
+		return err
+	}
+
+	// TODO: validate structure size against sector-size; ubuntu-image uses
+	// a tmp file to find out the default sector size of the device the tmp
+	// file is created on
+	return nil
+}
+
+func validateStructureType(s string, vol *GadgetVolume) error {
+	// Type can be one of:
+	// - "mbr" (backwards compatible)
+	// - "bare"
+	// - [0-9A-Z]{2} - MBR type
+	// - GPT UUID
+	// - hybrid ID
+	//
+	// Hybrid ID is 2 hex digits of MBR type, followed by 36 GUUID
+	// example: EF,C12A7328-F81F-11D2-BA4B-00A0C93EC93B
+
+	schema := vol.Schema
+	if schema == "" {
+		schema = "gpt"
+	}
+
+	if s == "" {
+		return errors.New(`type is not specified`)
+	}
+
+	if s == "bare" {
+		// unknonwn blob
+		return nil
+	}
+
+	if s == "mbr" {
+		// backward compatibility for type: mbr
+		return nil
+	}
+
+	var isGPT, isMBR bool
+
+	idx := strings.IndexRune(s, ',')
+	if idx == -1 {
+		// just ID
+		switch {
+		case validTypeID.MatchString(s):
+			isMBR = true
+		case validGUUID.MatchString(s):
+			isGPT = true
+		default:
+			return fmt.Errorf("invalid format")
+		}
+	} else {
+		// hybrid ID
+		code := s[:idx]
+		guid := s[idx+1:]
+		if len(code) != 2 || len(guid) != 36 || !validTypeID.MatchString(code) || !validGUUID.MatchString(guid) {
+			return fmt.Errorf("invalid format of hybrid type")
+		}
+	}
+
+	if schema != "gpt" && isGPT {
+		// type: <uuid> is only valid for GPT volumes
+		return fmt.Errorf("GUID structure type with non-GPT schema %q", vol.Schema)
+	}
+	if schema != "mbr" && isMBR {
+		return fmt.Errorf("MBR structure type with non-MBR schema %q", vol.Schema)
+	}
+
+	return nil
+}
+
+func validateRole(vs *VolumeStructure, vol *GadgetVolume) error {
+	if vs.Type == "bare" {
+		if vs.Role != "" && vs.Role != "mbr" {
+			return fmt.Errorf("conflicting type: %q", vs.Type)
+		}
+	}
+	vsRole := vs.Role
+	if vs.Type == "mbr" {
+		// backward compatibility
+		vsRole = "mbr"
+	}
+
+	switch vsRole {
+	case "system-data":
+		if vs.Label != "" && vs.Label != "writable" {
+			return fmt.Errorf(`role of this kind must have an implicit label or "writable", not %q`, vs.Label)
+		}
+	case "mbr":
+		if vs.Size > 446 {
```

Comment:
Can you add a reference that documents the size limit.

Reply:
I'll drop it in the description of structure fields